aiida/engine/processes/workchains/restart.py

```diff
@@ -0,0 +1,425 @@
+            message='The calculation failed for an unknown reason, twice in a row.')
+
+    def setup(self):
+        """Initialize context variables that are used during the logical flow of the `BaseRestartWorkChain`."""
+        self.ctx.calc_name = self._calculation_class.__name__
+        self.ctx.unhandled_failure = False
+        self.ctx.restart_calc = None
+        self.ctx.is_finished = False
+        self.ctx.iteration = 0
+
+    def should_run_calculation(self):
+        """Return whether a new calculation should be run.
+
+        This is the case as long as the last calculation has not finished successfully and the maximum number of
+        restarts has not yet been exceeded.
+        """
+        return not self.ctx.is_finished and self.ctx.iteration < self.inputs.max_iterations.value
+
+    def run_calculation(self):
+        """Run the next calculation, taking the input dictionary from the context at `self.ctx.inputs`."""
+        self.ctx.iteration += 1
+
+        try:
+            unwrapped_inputs = self.ctx.inputs
+        except AttributeError:
+            raise AttributeError('no calculation input dictionary was defined in `self.ctx.inputs`')
+
+        # Set the `CALL` link label
+        unwrapped_inputs['metadata']['call_link_label'] = 'iteration_{:02d}'.format(self.ctx.iteration)
+
+        inputs = self._wrap_bare_dict_inputs(self._calculation_class.spec().inputs, unwrapped_inputs)
+        calculation = self.submit(self._calculation_class, **inputs)
+
+        # Add a new empty list to the `errors_handled` extra. If any errors handled registered through the
+        # `register_error_handler` decorator return an `ErrorHandlerReport`, their name will be appended to that list.
+        errors_handled = self.node.get_extra('errors_handled', [])
+        errors_handled.append([])
+        self.node.set_extra('errors_handled', errors_handled)
+
+        self.report('launching {}<{}> iteration #{}'.format(self.ctx.calc_name, calculation.pk, self.ctx.iteration))
+
+        return ToContext(calculations=append_(calculation))
+
+    def inspect_calculation(self):
+        """Analyse the results of the previous calculation and call the error handlers when necessary."""
+        calculation = self.ctx.calculations[self.ctx.iteration - 1]
+
+        if calculation.is_excepted:
+            return self.exit_codes.ERROR_SUB_PROCESS_EXCEPTED  # pylint: disable=no-member
+
+        if calculation.is_killed:
+            return self.exit_codes.ERROR_SUB_PROCESS_KILLED  # pylint: disable=no-member
+
+        if calculation.is_finished_ok:
+
+            # The calculation was successful so we reset the unhandled failure flag
+            self.ctx.unhandled_failure = False
+
+            exit_code = self._call_sanity_checks(calculation)
+
+            if isinstance(exit_code, ExitCode):
+                if exit_code.status == 0:
+                    self.report('{}<{}> finished successfully, but sanity check failed, restarting'.format(
```

Comment:
I am a bit puzzled here. Shouldn't it be the opposite? I thought `exit_code.status == 0:` means sucesfull termination.